### DISCUSSION TITLE
I can't download videos from dyntube

### DISCUSSION THREAD

**Discussion by Jhomarers** [aclap-dev/video-downloadhelper#957] [Help and Q&A]
### Your question

I can't download videos from dyntube with VideoDownloadHelper, I wanted to know if the application can download these types of protected videos so I can buy it.  Here's the link: https://iframe.mediadelivery.net/embed/62736/be9ff4e4-eeeb-472b-b07a-97cbb77ed0d9

Windows 11 Home 64-Bit version 23H2 22631.3737, Chrome 126.0.6478.127 64-bit, not licensed VDH 9.0.2.12,  CoApp 2.0.19

**Comment by GrampaWildWilly** [comment]
I do not work for VDH.  I'm just another user.

Your problem report exhibits some deficiencies.  On the page where you created your problem report, there are instructions you ignored.  You should follow the instructions.  Or you can shorten what you post here by following this example:

Windows 11 Pro 64-bit version 23H2 build 22631.3810, Firefox 127.0.2 64-bit, licensed VDH 9.0.2.12a1 beta, CoApp 2.0.19

In addition, if you want somebody else to see if your problem can be replicated, you need to supply enough information to allow that other person to try it.  Telling me dyntube is just not enough.  I've never heard of the site & I'm not going to go look for it.  You need to make it EASY for me to help you.  Give me at least one URL of a web page on which a video is hosted that is causing you problems.  Don't say, "Pick one, any one will do."  No.  Like I said.  Make my life easy.  Give me specific URLs of web pages on which you are having problems.

**Comment by GrampaWildWilly** [comment]
Much better.

How did you even find this page?  When I visit it, I get this:

![#01](https://github.com/aclap-dev/video-downloadhelper/assets/75588964/4ce9b9fb-9a38-4f04-ab5a-3e2a6dde4815)

I don't know what's going on with that but it sure doesn't look promising.  403 is usually the code that means I don't have the credentials to access the data.  403 is usually accompanied with the warning `Forbidden`.

But that URL is not on a site named dyntube.  I thought you would post a URL on that site.  Maybe we can have better luck with something actually on dyntube.  Is dyntube a membership site?  If it is, I'm not going to be able to help you much.

**Comment by Mattmjs** [comment]
I don't get a 403, but it fails right away when attempting to download. Indeed it looks protected & I don't think VDH will
have a solution. To what I see and I could be wrong but it looks like they use encryption. You'll see in the image it uses a drmkey url,
three segments follow. The next drmkey url comes up  and another three segments follow after. Probably this pattern continues
all the way to the end of the video.


![dyntube](https://github.com/aclap-dev/video-downloadhelper/assets/160297904/6f76f8be-394c-4e67-a719-fecd82844716)

**Comment by sojusnik** [comment]
> I can't download videos from dyntube with VideoDownloadHelper, I wanted to know if the application can download these types of protected videos so I can buy it. Here's the link: https://iframe.mediadelivery.net/embed/62736/be9ff4e4-eeeb-472b-b07a-97cbb77ed0d9

It works here with v10.0.271.2, but [this](https://videos.dyntube.com/videos/exp-B7iJhDmH0urvkfoX2hXQ) one doesn't, any ideas why?

**Reply by Mattmjs** [reply]
Something must have changed some time after I looked at that video above. I'm not seeing what I saw in the screenshot.
It's different and it does start downloading. But your link I'm guessing is protected with encryption.
See the image taken from the terminal below, Unable to open key file.

<img width="652" height="287" alt="Screenshot from 2025-12-20 14-02-51" src="https://github.com/user-attachments/assets/86960cc7-9ebc-4810-88ae-eee5d57246d7" />

**Reply by GrampaWildWilly** [reply]
Yeah, I was able to download your first video as well but lordy lord it was slow.  Barely 500,000 bytes per second.  That's the speed of my old, lousy DSL connection.

For your second one, I checked in the Developer Tools Network Monitor.  There was an HLS master manifest.  But when I pushed it through ffprobe I got a whole lot of this:

```
[https @ 000001c2ffd12300] Opening 'https://cdn.dyntube.net/play/gbr-s/data/QxLVnAAH06uGsSPBa6BA/videos/OgJtCica066hCLG3aHHzw/v1/hls/720_A5DD1E82F671.m3u8?md5=ig7Hf-XnSgXtyI0uCUX0qg&expires=1766266453' for reading
[hls @ 000001c2ff7648c0] Opening 'https://api.dyntube.com/v1/live/videos/hls-key/18e48bfbd932f7d4/OgJtCica066hCLG3aHHzw' for reading
[https @ 000001c28172e040] HTTP error 403 Forbidden
[hls @ 000001c2ff7648c0] Unable to open key file https://api.dyntube.com/v1/live/videos/hls-key/18e48bfbd932f7d4/OgJtCica066hCLG3aHHzw, Server returned 403 Forbidden (access denied)
[hls @ 000001c2ff7648c0] Failed to open segment 0 of playlist 0
[hls @ 000001c2ff7648c0] Segment 0 of playlist 0 failed too many times, skipping
```

This might be what Matt posted but I can't read that.  The color scheme attacks my eyes & parts of what you posted are simply invisible to me.

Eventually, I think ffprobe just got tired & quit without generating a report.  Your web page does not show Firefox's usual DRM flag but I would call this stealth DRM.  There's a key file here & it's not being read because I don't have permission.  I might be tempted to say this is a subscription web site & if you log in, it might work.  Maybe.  That's a guess.  But it doesn't look like regular civilians are going to download this with any tool.

**Reply by Suwot** [reply]
this one is simple, the key url you see in hls is not the actual endpoint, it's the 'source' for internal player's logic to play with. original player's logic is deterministic, so it could be patched if devs wanted, but it's a hassle, and it definitely won't work in ANY DOWNLOADER unless player's logic is replicated before trying to decode segments

in other words, without special treatment nor VDH, nor other downloaders will download it, though the issue is definitely patchable

**Reply by sojusnik** [reply]
> in other words, without special treatment nor VDH, nor other downloaders will download it, though the issue is definitely patchable

Even tools like https://github.com/nilaoda/N_m3u8DL-RE/blob/main/README.en.md won't work on that? I've played around, but it seems that I can't identify the custom-hls-key, as mentioned [here](https://forum.videohelp.com/threads/413039-dyntube-claim-it-s-impossible-to-download-this-video-is-it).

**Reply by Suwot** [reply]
> > in other words, without special treatment nor VDH, nor other downloaders will download it, though the issue is definitely patchable
> 
> Even tools like https://github.com/nilaoda/N_m3u8DL-RE/blob/main/README.en.md won't work on that? I've played around, but it seems that I can't identify the custom-hls-key, as mentioned [here](https://forum.videohelp.com/threads/413039-dyntube-claim-it-s-impossible-to-download-this-video-is-it).

Well, as I said, if a developer of a downloader-tool didn't patch it specifically for this cdn/player, it won't work 'out of the box' – same situation for any downlaoder. the reason download is failing not the inability of a tool to download it, but the lack of proper playlist structure to download.... playlist itself has to be patched first, and to do so ANY tool has to replicate the player/cdn logic from this website, and only after that attempt to download it – this is why all tools will fail, unless they specifically patch logic for this website/cdn/player (your example from dyntube)

I personally never got a request to patch dyntube, so I suppose the website is just not popular enough for other tools to implement the patch...

**Reply by sojusnik** [reply]
I've managed to download [this](https://videos.dyntube.com/videos/exp-B7iJhDmH0urvkfoX2hXQ) video using [N_m3u8DL-RE](https://github.com/nilaoda/N_m3u8DL-RE). Maybe the following will help someone or the devs here:
1. [Download](https://github.com/nilaoda/N_m3u8DL-RE/releases) the binary, extract it to `~/.local/bin/` and make it executable
2. Open the site with the video, then the browser's network monitor, reload the site
3. Search in the network monitor for `m3u8`, something like `https://cdn.dyntube.net/play/gbr-s/data/QxLVnAAH06uGsSPBa6BA/videos/FvI0QlYXR0W0FpiCA5gsQ/v1/hls/720_A7D355910632.m3u8?md5=wRRlBRmPkyuDYqRuinnS7g&expires=1767319353` should appear, which you copy
4. Go to the response tab of the mentioned entry, where you should see something like `#EXT-X-KEY:METHOD=AES-128,URI="https://api.dyntube.com/v1/live/videos/hls-key/ceec0f3e9ffd7858/FvI0QlYXR0W0FpiCA5gsQ",IV=0x15ab5003bd6061c83cdfa3e53c553045`
5. Search for `hls-key` in the network monitor and you should find the above mentioned entry, go to its response tab where you'll find a code, something like `ZBdO0/lw0vTRucUPOLh00g==` which is your hls-key
6. Combine everything to: `N_m3u8DL-RE --custom-hls-key "ZBdO0/lw0vTRucUPOLh00g==" "https://cdn.dyntube.net/play/gbr-s/data/QxLVnAAH06uGsSPBa6BA/videos/FvI0QlYXR0W0FpiCA5gsQ/v1/hls/720_A7D355910632.m3u8?md5=wRRlBRmPkyuDYqRuinnS7g&expires=1767319353"` and execute in the terminal, which should download a mp4 file
7. Enjoy!

**Reply by Suwot** [reply]
manual operations always beat automated patches in reliability, the tool you described allows to overwrite key from manifest to your custom key (exactly the reason why VDH failed), so you got a proper media downloaded after you've found correct key

this will work not only here, but on other websites with this issue – because it's a manual approach

**Reply by Mattmjs** [reply]
I tried as a test to use ffmpeg, but failed. It's seems VDH will never work on cases like this.

What an AI said about it below

The error you're seeing ("HTTP error 403 Forbidden" followed by "Unable to open key file") happens because FFmpeg is still trying to fetch the key from the remote URI listed in the #EXT-X-KEY tag inside the .m3u8 playlist — even when you use -decryption_key.
Unfortunately, FFmpeg's -decryption_key option does not override or ignore the key URI in the playlist. It only works in specific cases (like single encrypted MP4 files or certain configurations), but for standard AES-128 HLS streams, FFmpeg always attempts to open the key URI first.

Recommended workaround: Edit the .m3u8 playlist to use a local key file

1. Download the master/variant .m3u8 file (and any sub-playlists if it's a multi-bitrate stream).
2. Open the playlist(s) in a text editor.
3. Find the line like: #EXT-X-KEY:METHOD=AES-128,URI="https://example.com/somekey.key",IV=0x...
4. Create a local binary key file (not hex text):
- Use this command (replace with your 32-character hex key): echo YOUR32CHARHEXKEY | xxd -r -p > local.key

Example: echo abcdef1234567890abcdef1234567890 | xxd -r -p > local.key
This creates a 16-byte binary file named local.key.

5. Change the URI in the playlist to point to your local key file:
#EXT-X-KEY:METHOD=AES-128,URI="local.key",IV=0x...
(Use a relative path like "local.key" if the key file is in the same folder as the edited .m3u8.)
6. Save the edited playlist as something like modified.m3u8.
7. Run FFmpeg on the modified local playlist: ffmpeg -allowed_extensions ALL -i modified.m3u8 -c copy output.mp4

- Make sure local.key and modified.m3u8 are in the same directory.
- The -allowed_extensions ALL flag avoids security blocks on .key files.
- Add -protocol_whitelist file,http,https,tcp,tls,crypto if needed for mixed local/remote segments.

This way, FFmpeg loads the key locally and skips the remote URI entirely.

**Reply by GrampaWildWilly** [reply]
That is hilarious.  You're basically spoofing a key.  That means that anybody who tries to put an encryption key on their content can be easily thwarted like this.  I'll have to try it next time I encounter such a case.

If you're going to save an HLS stream manifest (not a master for this case) to your system, it might be necessary to edit every URL in the file.  Granted, this would be a global replace, but it would still need to be done.  If the URLs of the individual chunks don't start with http:// or https://, then you've got relative URLs & they won't work as is.  You have to prepend the web site's directory structure to each URL in the manifest.  You should be able to grab the URL of the manifest itself & prepend some part of that to each URL line inside the manifest you've copied to your system.

Thanks for the xxd idea.  I'd never heard of that tool before.  Now I have it safely squirreled away for use in that hypothetical future case.

**Reply by Suwot** [reply]
> That means that anybody who tries to put an encryption key on their content can be easily thwarted like this.

In many cases you won't be able to see a key in Network in DevTools, because for proper protection, sensitive data is fetched by player's internal JS logic, which is not exposed in Network calls. Dyntube case is just a simple one.

**Reply by Suwot** [reply]
> I tried as a test to use ffmpeg, but failed. It's seems VDH will never work on cases like this.

VDH can patch for dyntube, it's just not worth it for 1 user (who already got the solution) when there are bigger problems on the TODO list

what AI said is a manual nightmare, because even in a simple case like dyntube the segment paths are dynamic, meaning FFmpeg will resolve them upon your local file and will fail even with correct key. FFmpeg original build doesn't have an argument to accept custom key for decription (unless it's a DRM key, definitely not AES-128 which dyntube uses), so the only way to handle it via FFmpeg is what AI suggested – local patched manifest, and it's way easier with the tool [sojusnik](https://github.com/sojusnik) mentioned than with FFmpeg

**Comment by GrampaWildWilly** [comment]
I was going to post a whole long narrative of how I managed to use ffmpeg to download the video in the original poster's initial post at the top of this thread.  But I failed miserably.  It seems that you can't just make up a key file.  You have to know the actual key value & I could not find that in the information provided by the Firefox Dev Tools.  Maybe somebody else can get this to work.  I can't.  It's not quite as simple as the AI response @Mattmjs Matt got that he posted above.  Keys really are required & the security is not as lax as I originally concluded.  The key file actually does appear in the Network Monitor.  I tried just saving it on my system & using it in ffmpeg.  But it didn't work.  There must be some additional manipulation you have to perform on the key & I don't know what that is.  Live & learn.

**Reply by Mattmjs** [reply]
I don't think it was quite accurate what it provided the first time. I was direct and more on point second time around.
You can see my query and it's response below with new information that I didn't get before. With this new information
the key had to be converted. After that I used the new command in ffmpeg, failure. I thought I was too slow with the expiry
in the link. Refreshed the page to get a new link, quickly running command but failure again. I haven't tried the next option
with local file with the updated information it provided. 

My query

The playlist points to this link https://api.dyntube.com/v1/live/videos/player-hls-key/wzHHa3GLCh660aciCtJgO/18e48bfbd932f7d4/OgJtCica066hCLG3aHHzw and the response payload for this link contains this AMhgJT5nwNDyX138wt36Pg== which I assume is the decryption key. So if this is the decryption key what is the process to use ffmpeg. I can't use -decryption_key because it still fails, how could it be done with a local file.

AI response

The provided payload "AMhgJT5nwNDyX138wt36Pg==" is a base64-encoded 16-byte AES-128 decryption key. Decoded to hexadecimal format (which FFmpeg requires for the -decryption_key option), it is 00c860253e67c0d0f25f5dfcc2ddfa3e.

Why -decryption_key Might Have Failed Previously  ( **Exactly what I did first time around** )

- If you passed the base64 string directly (e.g., -decryption_key AMhgJT5nwNDyX138wt36Pg==), it would fail because FFmpeg expects a 32-character hexadecimal string, not base64.

- Other common issues: Mismatched IV (initialization vector) handling, expired/invalid tokens in the playlist URL preventing segment access, or missing HTTP headers (e.g., Referer, User-Agent) if the CDN enforces them.

- Assuming the token in your playlist URL (?md5=os3dwJDSzGhsHuIdQzHlHw&expires=1767538668) is still valid and segments are accessible, try the corrected command below. ( **As I said it still fails with the updated information below** )

Using FFmpeg with -decryption_key (Simplest Online Method)
This overrides the key URI in the playlist and lets FFmpeg fetch/decrypt segments directly from the CDN.

ffmpeg -decryption_key 00c860253e67c0d0f25f5dfcc2ddfa3e \
       -headers "User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36\r\nReferer: https://the-original-site.com/\r\nOrigin: https://the-original-site.com/\r\n" \
       -i "https://cdn.dyntube.net/play/gbr-s/data/QxLVnAAH06uGsSPBa6BA/videos/OgJtCica066hCLG3aHHzw/v1/hls/720_A5DD1E82F671.m3u8?md5=os3dwJDSzGhsHuIdQzHlHw&expires=1767538668" \
       -c copy -bsf:a aac_adtstoasc output.mp4

- Replace the Referer/Origin with the actual site where the video is embedded (check browser dev tools).

- Add other headers like Cookie if required for authentication.

- If IV is explicitly set in the playlist and causes issues, FFmpeg typically handles it automatically (using segment sequence as IV if unspecified).

- Debug tip: Run with -v debug to see errors (e.g., "Unable to open key file" or segment access denied).

**Local method**
Using FFmpeg with a Local Key File (Offline-Friendly Method)  ( **Haven't tried with updated information** )
If the above fails due to key fetching or you prefer local handling, save the key locally and modify the playlist to reference it. This works well for DynTube streams where the key is fetched separately.

1. Save the key as a binary file:
- Use Python (or an online base64-to-binary decoder):

import base64
with open('key.bin', 'wb') as f:
    f.write(base64.b64decode('AMhgJT5nwNDyX138wt36Pg=='))

- This creates key.bin (16 bytes binary).

2. Download the playlist:

curl -H "User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36" \
     -H "Referer: https://the-original-site.com/" \
     "https://cdn.dyntube.net/play/gbr-s/data/QxLVnAAH06uGsSPBa6BA/videos/OgJtCica066hCLG3aHHzw/v1/hls/720_A5DD1E82F671.m3u8?md5=os3dwJDSzGhsHuIdQzHlHw&expires=1767538668" \
     -o playlist.m3u8

- Adjust headers as needed.

3. Edit the playlist (playlist.m3u8):
- Open in a text editor.

- Find the #EXT-X-KEY line (it should look like #EXT-X-KEY:METHOD=AES-128,URI="https://api.dyntube.com/v1/live/videos/player-hls-key/wzHHa3GLCh660aciCtJgO/18e48bfbd932f7d4/OgJtCica066hCLG3aHHzw",IV=...).

- Change the URI to point to your local key: URI="key.bin" (assuming key.bin is in the same directory as playlist.m3u8). Keep the METHOD and IV unchanged.

- If segment URLs are relative (e.g., segment0.ts), ensure they're resolvable (FFmpeg will prepend the base URL from the original playlist location). If absolute, no issue.

- Save as modified_playlist.m3u8.

4. Run FFmpeg: ffmpeg -allowed_extensions ALL \
       -headers "User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36\r\nReferer: https://the-original-site.com/\r\n" \
       -i modified_playlist.m3u8 \
       -c copy -bsf:a aac_adtstoasc output.mp4

- -allowed_extensions ALL: Allows non-standard extensions if needed.

- FFmpeg will fetch segments online (using headers for access) but use your local key.bin for decryption.

- If you want fully offline: Download all .ts segments listed in the playlist (using curl/wget with similar headers/token), place them in the same directory, and ensure paths in the m3u8 are relative/local.

Additional Tips

- If Still Fails: The token might be expired (expires=1767538668 is a Unix timestamp ~Jan 5, 2026; regenerate a fresh playlist URL if needed). Or, segments require the same md5/expires query params—append them manually to segment URLs in the edited m3u8 if missing. 

- Alternative Tool: If FFmpeg struggles, use N_m3u8DL-RE (better for protected HLS): ( **By far the easier option for this type of video** )

N_m3u8DL-RE "PASTE_FRESH_M3U8_URL_HERE" \
            --custom-hls-key "00c860253e67c0d0f25f5dfcc2ddfa3e" \
            --header "User-Agent: Mozilla/5.0 ..." \
            --header "Referer: https://original-site.com/" \
            --save-dir . --save-name video_720p

**Reply by GrampaWildWilly** [reply]
I saved the HLS stream manifest on my system & did a global change to make all the relative URLs in there into absolute URLs.  I changed the reference to the key file to be the key file I saved to my system.  I did not use any parameter of ffmpeg to do key manipulation.  I left the pointer to the key file in the stream manifest.  That is the other way this is supposed to work.  Ffmpeg just showed every single chunk of the download getting permission errors.  I don't have Python, curl, or any of those other Linuxy type tools & I'm not sufficiently motivated to get them.  I have admitted defeat.  I am moving on.
